### PR TITLE
Mosviz redshift slider

### PR DESCRIPTION
**Pull request by kecnry**
### Description

This pull request is an extension of #961 and adds a redshift slider in Mosviz along with new `mosviz.get_spectrum_1d(row=None)` and `mosviz.get_spectrum_2d(row=None)` (defaults to the currently selected row if `row` is not provided) helper methods to access the underlying data products with their respective redshifts applied.

Note: support for RVs in the slider dropdown is supported, but roundtripping is not perfect.  Support for an RV column in the table connected to the slider is not (currently) included.

This also currently includes #980 (which would preferably be merged first and then this PR rebased onto main)

Wishlist:
- [x] fix "roundtripping" between redshift and RV (if possible)
    * roundtripping internally is confirmed to be fine.  Issue is that the _step_ of the slider itself is forcing values to round when switching between redshift and RV.  Fixing this would require a redesign of how the bounds/step of the slider are determined.  Determined to be out-of-scope (same issue existed before for specviz, just wasn't apparent since the redshift value isn't shown and it generally "rounds back" to the original value when re-selecting redshift) and filed as #987.
- [x] investigate behavior for hidden redshift column when changing slider (any changed data will automatically force that column to be shown which conflicts with the `hide_column` logic).  
    * postponing to wait for user feedback (some voted that the current behavior might actually be preferred).  Changing this will likely require storing the desired list internally and re-setting the visible columns, or overriding the behavior internal to glue.
- [x] ~if RV column is within scope: design and implement~
- [x] try to minimize/remove redshift column "flickering" when changing rows.  (Now fairly minimal but not gone.  Fixing this completely is likely a larger fix that can also cover the flickering in the viewers when changing rows and is likely out of scope)
- [x] add test coverage. (Some done, but might be able to be improved before merge)
- [x] merge #980 and rebase

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

**Comment by pllim** [comment]
> Issue is that the step of the slider itself is forcing values to round when switching between redshift and RV.

Sounds like we should open a follow-up issue about this. Have you opened a follow-up issue yet?

**Review by pllim — COMMENTED** [review]
Seems to work but can be improved with the following:

1. Add some user-facing documentation on the new UI improvement and also the two new public methods mentioned in change log.
2. When calling `mosviz.get_spectrum_1d()` and `mosviz.get_spectrum_2d()`, I see a warning that says `WARNING:root:Warning: Applying the value from the redshift slider to the output spectra. To avoid seeing this warning, explicitly set the apply_slider_redshift argument to True or False.`, which is very confusing because I am calling the methods using default parameters, which are usually the recommended ones. Suggest change the default into something that does not produce a warning (unless this is what PO wanted...).
3. Is blueshift not supported at all? Don't some lines get blueshifted in some scientific situations?

**Comment by kecnry** [comment]
> 1. Add some user-facing documentation on the new UI improvement and also the two new public methods mentioned in change log.

Added basic information (and link to the Specviz redshift docs) in 8470c01.  Let me know if you think anything is missing that needs to be there.

> 2. When calling `mosviz.get_spectrum_1d()` and `mosviz.get_spectrum_2d()`, I see a warning that says `WARNING:root:Warning: Applying the value from the redshift slider to the output spectra. To avoid seeing this warning, explicitly set the apply_slider_redshift argument to True or False.`, which is very confusing because I am calling the methods using default parameters, which are usually the recommended ones. Suggest change the default into something that does not produce a warning (unless this is what PO wanted...).

This is to mirror the behavior of `specviz.get_spectra()` and was added in 8b8f2b0fc3ffa1d01e6154068399300ad7447ec0 by @rosteen but I don't know the motivation for that decision.  If we make any changes, I think they should be made in both cases for consistency.

> 3. Is blueshift not supported at all? Don't some lines get blueshifted in some scientific situations?

Blueshift should be supported, but will need to currently be entered manually as otherwise the lower limit of the slider is always zero/positive.  This case should be a consideration if/when redesigning the slider (relevant to #987).

File: docs/mosviz/redshift.rst
Comment (pllim) [inline docs/mosviz/redshift.rst:11]:
```suggestion
    lines using, e.g., the "Hide All" button in the line lists UI.
```

File: docs/mosviz/redshift.rst
Comment (pllim) [inline docs/mosviz/redshift.rst:24]:
```suggestion
:meth:`~jdaviz.configs.mosviz.helper.Mosviz.get_spectrum_1d` and :meth:`~jdaviz.configs.mosviz.helper.Mosviz.get_spectrum_2d`,
```

File: docs/mosviz/redshift.rst
Comment (pllim) [inline docs/mosviz/redshift.rst:21]:
```suggestion
using :meth:`~jdaviz.configs.mosviz.helper.Mosviz.update_column`.
```

**Comment by pllim** [comment]
I'll approve when tests are added. 😉

**Review by rosteen — APPROVED** [review]
Looks good, everything seems to be behaving and performing well. The only thing I think might trip some people up is if they wanted to set the redshift for all the rows - it might be nice to have an option to set `row='all'` in the `mosviz.update_column` call to do that without needing to construct an array of the correct length. But I don't think that should hold up the PR, so I'm going to approve and if you want to implement that now I can take another look before merging.

**Comment by rosteen** [comment]
@pllim There are tests now, do you approve? 😁

File: jdaviz/configs/mosviz/tests/test_helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_helper.py:368]: If redshift is a Quantity, you can use `assert_quantity_allclose` from `astropy`.

Comment (pllim) [reply jdaviz/configs/mosviz/tests/test_helper.py:368]: Otherwise, `numpy.testing.assert_allclose` is more consistent with the rest of the test suite.

File: jdaviz/configs/mosviz/tests/test_helper.py
Comment (pllim) [inline jdaviz/configs/mosviz/tests/test_helper.py:369]: Is there a reason to just compare the name instead of using `isinstance`?

**Review by pllim — APPROVED** [review]

**Comment by pllim** [comment]
Thanks!

**Pull request merged**
